# Example to gather basic statistics on world population

Preparation to run this analysis
* Upload data to Google Drive
* Download the notebook from Github classroom to your computer and upload it to Colab
* Select a compute resource
YOU ARE NOW ALL SET TO RUN THE ANALYSIS!



#### Define problem statement

**Problem statement**

I am a United Nations policy maker that wants basic statistics about the world population in the latest data we have available. This information will help me with recommendations for resource allocation.

Notes:
When the objectives of the task are clear, the analyst can do an efficient job of extracting the relevant data.




**Disclaimer:**

In this workshop, the analysis that we provide below is not exhaustive. Please use this as a guide to learn the concepts of data analysis that were taught in the course (problem statement, data collection, cleaning and analysis). Here we have used data from the latest available date. In a real scenario, the analyst would want to look at more recent data and/or apply various other math/statistical methods to help answer their question.

For the purposes of this course, let us assume that the year is 2021

*Instructions*
* We will be using Python to run the code
* **We want to focus on the concepts of analyzing data**
    * If you have prior experience in Python, please feel free to try additional questions - as we walk through the course
    * If you are new to Python, you will not be asked to write Python code. The code is provided. A novice usually finds it difficult to analyze the data as well as follow the code. We re-iterate that you focus on the concepts (and the outcome of the analysis) rather than the code. At your liesure, we encourage you to implement the analysis in your tool of choice (such as Excel or other.)
* For those interested to pursue a career path in data analysis, the code we present below is higher than an introductory level Python course. We present popular tools and code patterns in the hope that they will help you accelerate your learning journey.

*About this course*
* The purpose of this data analysis is to understand basic trends in World Population. We want to explore the data that will help answer the questions - What, Where or When. In this course, we will focus on introductory data analysis.
* Advanced data analysis (not covered here) would ask questions on the "Why" and also help come up with forecast for the future.
* Any data analysis always starts with the basic questions.

Data sources:
United Nations, World Population Prospects (2022)


import packages

In [ ]:
import pandas as pd
import numpy as np


#### Select data

In [ ]:
# Import data into Colab by clicking on the "Files" icon and then clicking the "Upload" button. Select the CSV from your computer.

# pd.read_csv(""/content/drive/MyDrive/population-and-demography.csv)

In [ ]:
# population data contains population for countries,  several groups of countries and the world
pg_data = pd.read_csv("/content/population-and-demography.csv")
pg_data.head()

Country name  Year  Population  Population of children under the age of 1  \
0  Afghanistan  1950     7480464                                   301735.0   
1  Afghanistan  1951     7571542                                   299368.0   
2  Afghanistan  1952     7667534                                   305393.0   
3  Afghanistan  1953     7764549                                   311574.0   
4  Afghanistan  1954     7864289                                   317584.0   

   Population of children under the age of 5  \
0                                    1248282   
1                                    1246857   
2                                    1248220   
3                                    1254725   
4                                    1267817   

   Population of children under the age of 15  Population under the age of 25  \
0                                     3068855                         4494349   
1                                     3105444                         4552138   
2                                     3145070                         4613604   
3                                     3186382                         4676232   
4                                     3231060                         4741371   

   Population aged 15 to 64 years  Population older than 15 years  \
0                         4198587                         4411609   
1                         4250002                         4466098   
2                         4303436                         4522464   
3                         4356242                         4578167   
4                         4408474                         4633229   

   Population older than 18 years  ...  Population aged 15 to 19 years  \
0                         3946595  ...                          757113   
1                         3993640  ...                          768616   
2                         4041439  ...                          781411   
3                         4088379  ...                          794308   
4                         4136116  ...                          806216   

   Population aged 20 to 29 years  Population aged 30 to 39 years  \
0                         1241348                          909953   
1                         1260904                          922765   
2                         1280288                          935638   
3                         1298803                          948321   
4                         1316768                          961484   

   Population aged 40 to 49 years  Population aged 50 to 59 years  \
0                          661807                          467170   
1                          667015                          468881   
2                          672491                          470898   
3                          678064                          472969   
4                          684153                          475117   

   Population aged 60 to 69 years  Population aged 70 to 79 years  \
0                          271905                           92691   
1                          273286                           94358   
2                          274852                           96026   
3                          276577                           97705   
4                          278210                           99298   

   Population aged 80 to 89 years  Population aged 90 to 99 years  \
0                            9499                             123   
1                           10155                             118   
2                           10721                             139   
3                           11254                             166   
4                           11793                             190   

   Population older than 100 years  
0                              0.0  
1                              0.0  
2                              0.0  
3                              0.0  
4                              0.0  

[5 rows 

*In the real world, the raw data are ingested into the raw layer or the "Bronze layer"*

#### Clean data

##### Transform the datasets to meet our needs
* Normally, the data are merged with other data or are grouped to meet the analysis needs. Here in the first exercise, we will work on one data.
* The population data that we need is present in the ingested data. Further, we will be analyzing data by country.

##### Clean data and understand a summary of the data
*   Remove null values
*   Show a summary of the data


In [ ]:
# The number of rows before removing null values
print(f"Number of rows is {pg_data.shape[0]}")

Number of rows is 18288


In [ ]:
pg_data = pg_data.dropna()
# The number of rows after removing null values
print(f"Number of rows is {pg_data.shape[0]}")
pg_data.head()

Number of rows is 18288


Country name  Year  Population  Population of children under the age of 1  \
0  Afghanistan  1950     7480464                                   301735.0   
1  Afghanistan  1951     7571542                                   299368.0   
2  Afghanistan  1952     7667534                                   305393.0   
3  Afghanistan  1953     7764549                                   311574.0   
4  Afghanistan  1954     7864289                                   317584.0   

   Population of children under the age of 5  \
0                                    1248282   
1                                    1246857   
2                                    1248220   
3                                    1254725   
4                                    1267817   

   Population of children under the age of 15  Population under the age of 25  \
0                                     3068855                         4494349   
1                                     3105444                         4552138   
2                                     3145070                         4613604   
3                                     3186382                         4676232   
4                                     3231060                         4741371   

   Population aged 15 to 64 years  Population older than 15 years  \
0                         4198587                         4411609   
1                         4250002                         4466098   
2                         4303436                         4522464   
3                         4356242                         4578167   
4                         4408474                         4633229   

   Population older than 18 years  ...  Population aged 15 to 19 years  \
0                         3946595  ...                          757113   
1                         3993640  ...                          768616   
2                         4041439  ...                          781411   
3                         4088379  ...                          794308   
4                         4136116  ...                          806216   

   Population aged 20 to 29 years  Population aged 30 to 39 years  \
0                         1241348                          909953   
1                         1260904                          922765   
2                         1280288                          935638   
3                         1298803                          948321   
4                         1316768                          961484   

   Population aged 40 to 49 years  Population aged 50 to 59 years  \
0                          661807                          467170   
1                          667015                          468881   
2                          672491                          470898   
3                          678064                          472969   
4                          684153                          475117   

   Population aged 60 to 69 years  Population aged 70 to 79 years  \
0                          271905                           92691   
1                          273286                           94358   
2                          274852                           96026   
3                          276577                           97705   
4                          278210                           99298   

   Population aged 80 to 89 years  Population aged 90 to 99 years  \
0                            9499                             123   
1                           10155                             118   
2                           10721                             139   
3                           11254                             166   
4                           11793                             190   

   Population older than 100 years  
0                              0.0  
1                              0.0  
2                              0.0  
3                              0.0  
4                              0.0  

[5 rows 

Looks like no null values were present

*Tell me about this data*

In [ ]:
pg_data.describe()

Year    Population  Population of children under the age of 1  \
count  18288.000000  1.828800e+04                               1.828800e+04   
mean    1985.500000  1.264704e+08                               3.133497e+06   
std       20.783173  5.888512e+08                               1.416701e+07   
min     1950.000000  1.363000e+03                               2.500000e+01   
25%     1967.750000  2.915915e+05                               6.663750e+03   
50%     1985.500000  3.833998e+06                               8.835200e+04   
75%     2003.250000  1.678546e+07                               4.630005e+05   
max     2021.000000  7.909295e+09                               1.397837e+08   

       Population of children under the age of 5  \
count                               1.828800e+04   
mean                                1.482571e+07   
std                                 6.738437e+07   
min                                 1.360000e+02   
25%                                 3.199525e+04   
50%                                 4.237845e+05   
75%                                 2.160046e+06   
max                                 6.903607e+08   

       Population of children under the age of 15  \
count                                1.828800e+04   
mean                                 4.109523e+07   
std                                  1.884170e+08   
min                                  4.160000e+02   
25%                                  8.954150e+04   
50%                                  1.186122e+06   
75%                                  5.905945e+06   
max                                  2.015023e+09   

       Population under the age of 25  Population aged 15 to 64 years  \
count                    1.828800e+04                    1.828800e+04   
mean                     6.376260e+07                    7.742951e+07   
std                      2.942519e+08                    3.676519e+08   
min                      6.230000e+02                    7.480000e+02   
25%                      1.395415e+05                    1.702635e+05   
50%                      1.843100e+06                    2.246772e+06   
75%                      9.025130e+06                    9.641250e+06   
max                      3.239281e+09                    5.132999e+09   

       Population older than 15 years  Population older than 18 years  \
count                    1.828800e+04                    1.828800e+04   
mean                     8.537210e+07                    7.819624e+07   
std                      4.048669e+08                    3.720170e+08   
min                      8.490000e+02                    7.520000e+02   
25%                      1.867160e+05                    1.664175e+05   
50%                      2.482104e+06                    2.238130e+06   
75%                      1.035435e+07                    9.239904e+06   
max                      5.893679e+09                    5.516283e+09   

       Population at age 1  ...  Population aged 15 to 19 years  \
count         1.828800e+04  ...                    1.828800e+04   
mean          3.011213e+06  ...                    1.178226e+07   
std           1.366200e+07  ...                    5.512604e+07   
min           2.600000e+01  ...                    1.100000e+02   
25%           6.473750e+03  ...                    2.629650e+04   
50%           8.582400e+04  ...                    3.369695e+05   
75%           4.407875e+05  ...                    1.626211e+06   
max           1.384787e+08  ...                    6.235761e+08   

       Population aged 20 to 29 years  Population aged 30 to 39 years  \
count                    1.828800e+04                    1.828800e+04   
mean                     2.087288e+07                    1.715870e+07   
std                      9.886099e+07                    8.240460e+07   
min                      1.580000e+02                    1.370000e+02   
25%                      4.505075e+04                    3.660

Notes on the data:
* Maximum population is about 7.9 billion. **This seems odd**
* Data is for each country and given by year.
* Total population and population by age groups are provided

Question:
* What are the minimum, maximum and mean population in 2021?




*In the real world, the transformed (and cleaned) data are stored into the intermediate layer or the "Silver layer". This data is now ready for analytics use.*
*   Additional data cleaning, formatting are done to ensure the data types of each column are making sense.
    * This can include checks if country names are consistent.
    * Also, include data aggregation.e.g., represent minute data as hourly. For this data, we do not need to do aggregation.

##### Perform additional checks to make sure the data looks alright and do additional cleaning if necessary

*   *Show countries that have greater than China's 2021 population*

Please note the coding pattern
* We will define a condition that we need (see variable `condition`)
* Then we apply that condition on the dataframe

In [ ]:
condition = pg_data["Population"]>1425893500
pg_data[condition].head() # find me data whose population is more than china

Country name  Year  Population  Population of children under the age of 1  \
866    Asia (UN)  1952  1435002800                                 53441104.0   
867    Asia (UN)  1953  1466384300                                 55347424.0   
868    Asia (UN)  1954  1498736600                                 56284908.0   
869    Asia (UN)  1955  1532902800                                 57757990.0   
870    Asia (UN)  1956  1567477800                                 58110908.0   

     Population of children under the age of 5  \
866                                  224022800   
867                                  234208110   
868                                  243142940   
869                                  251023420   
870                                  257773180   

     Population of children under the age of 15  \
866                                   538876300   
867                                   555507650   
868                                   573178050   
869                                   592322200   
870                                   611292300   

     Population under the age of 25  Population aged 15 to 64 years  \
866                       807658750                       838080200   
867                       828502800                       852104060   
868                       849711040                       865980900   
869                       872123260                       880300600   
870                       894578940                       895319230   

     Population older than 15 years  Population older than 18 years  ...  \
866                       896121900                       809027500  ...   
867                       910872000                       823001860  ...   
868                       925554240                       837504260  ...   
869                       940576830                       852483800  ...   
870                       956182000                       867582140  ...   

     Population aged 15 to 19 years  Population aged 20 to 29 years  \
866                       142534500                       236827220   
867                       144235760                       241922620   
868                       145128240                       246899860   
869                       145856960                       251657020   
870                       146785280                       256619490   

     Population aged 30 to 39 years  Population aged 40 to 49 years  \
866                       181836320                       142478320   
867                       184647220                       144652830   
868                       187728160                       147067150   
869                       191298400                       149476320   
870                       195036580                       151727950   

     Population aged 50 to 59 years  Population aged 60 to 69 years  \
866                        99356310                        60423012   
867                       101525256                        61161500   
868                       103757150                        62098856   
869                       105985730                        63177212   
870                       108235304                        64280350   

     Population aged 70 to 79 years  Population aged 80 to 89 years  \
866                        26108144                         6141401   
867                        26221600                         6093195   
868                        26404860                         6062275   
869                        26682296                         6038162   
870                        27076954                         6021905   

     Population aged 90 to 99 years  Population older than 100 years  
866                          416713                           4511.0  
867                          412023                           4633.0  
868                          407669                           4325.0  
869                          4047

In [ ]:
# Some values do not correspond to country. Let us remove these entries from the data.

# get all values for country names whose population is more than 1.47 billion
condition = pg_data["Population"]>1425893500
remove_words = list(pg_data[condition]["Country name"].unique())
remove_words

['Asia (UN)',
 'Less developed regions',
 'Less developed regions, excluding China',
 'Less developed regions, excluding least developed countries',
 'Lower-middle-income countries',
 'Upper-middle-income countries',
 'World']

*   *Remove "World" and other non-country entities from the Country name column*
    * Count number of countries

In [ ]:
# We can generalize the above words to maximize cleaning
remove_words = [ 'regions',
 'countries',
 'World','UN','SIDS']

In [ ]:
# remove regions
pg_data_nona_clean = pg_data.copy()
def remove_non_countries(remove_words,pg_data_nona_clean):
    for word in remove_words:

        mask = ~(pg_data_nona_clean["Country name"].str.contains(word))
        pg_data_nona_clean = pg_data_nona_clean\
            [mask]

        # num of unique countries in the data
        num_of_countries = len(pg_data_nona_clean["Country name"].unique())
        print("Number of countries in the data \
            are {}".format(num_of_countries )
        )
        # print(word)
    return pg_data_nona_clean
pg_data_nona_clean = remove_non_countries(remove_words,pg_data_nona_clean)


Number of countries in the data             are 250
Number of countries in the data             are 244
Number of countries in the data             are 243
Number of countries in the data             are 237
Number of countries in the data             are 236


Still, there are way too many countries. Let us see if non-countries exist in the lower population.

#### Analyze data

* During analysis also, data clean may be required.
* As we discover patterns, we may need to perform additional cleaning.

In [ ]:
# countries with population below 100,000 in 2021
condition1 = (pg_data_nona_clean["Population"]<=100000)
condition2 = (pg_data_nona_clean["Year"]==2021)
pg_data_nona_clean[condition1 & condition2]["Country name"].unique()

array(['American Samoa', 'Andorra', 'Anguilla', 'Antigua and Barbuda',
       'Bermuda', 'Bonaire Sint Eustatius and Saba',
       'British Virgin Islands', 'Cayman Islands', 'Cook Islands',
       'Dominica', 'Falkland Islands', 'Faroe Islands', 'Gibraltar',
       'Greenland', 'Guernsey', 'Isle of Man', 'Liechtenstein',
       'Marshall Islands', 'Monaco', 'Montserrat', 'Nauru', 'Niue',
       'Northern Mariana Islands', 'Palau', 'Saint Barthelemy',
       'Saint Helena', 'Saint Kitts and Nevis',
       'Saint Martin (French part)', 'Saint Pierre and Miquelon',
       'San Marino', 'Sint Maarten (Dutch part)', 'Tokelau',
       'Turks and Caicos Islands', 'Tuvalu', 'Wallis and Futuna'],
      dtype=object)

It looks like islands are included in the data as countries.

*Let us find out entities with minimum population and confirm if they are countries

In [ ]:
condition = pg_data_nona_clean["Year"]==2021
df = pg_data_nona_clean[condition].sort_values("Population").head(15)
df


Country name  Year  Population  \
16487                          Tokelau  2021        1869   
12095                             Niue  2021        1957   
5183                  Falkland Islands  2021        3786   
11087                       Montserrat  2021        4438   
13895                     Saint Helena  2021        5428   
14183        Saint Pierre and Miquelon  2021        5905   
13823                 Saint Barthelemy  2021       10888   
16991                           Tuvalu  2021       11229   
17927                Wallis and Futuna  2021       11654   
11519                            Nauru  2021       12533   
575                           Anguilla  2021       15779   
3527                      Cook Islands  2021       17028   
12743                            Palau  2021       18050   
2015   Bonaire Sint Eustatius and Saba  2021       26730   
2303            British Virgin Islands  2021       31145   

       Population of children under the age of 1  \
16487                                       35.0   
12095                                       27.0   
5183                                        43.0   
11087                                       41.0   
13895                                       42.0   
14183                                       45.0   
13823                                       87.0   
16991                                      259.0   
17927                                      138.0   
11519                                      341.0   
575                                        150.0   
3527                                       277.0   
12743                                      272.0   
2015                                       282.0   
2303                                       230.0   

       Population of children under the age of 5  \
16487                                        170   
12095                                        145   
5183                                         212   
11087                                        196   
13895                                        221   
14183                                        267   
13823                                        472   
16991                                       1250   
17927                                        753   
11519                                       1711   
575                                          800   
3527                                        1402   
12743                                       1308   
2015                                        1433   
2303                                        1164   

       Population of children under the age of 15  \
16487                                         545   
12095                                         514   
5183                                          672   
11087                                         595   
13895                                         758   
14183                                        1015   
13823                                        1454   
16991                                        3548   
17927                                        2714   
11519                                        4863   
575                                          2783   
3527                                         4029   
12743                                        3781   
2015                                         4543   
2303                                         4738   

       Population under the age of 25  Population aged 15 to 64 years  \
16487                             858                            1165   
12095                             720                            1148   
5183                             1083                            2703   
11087                            1243                            3077   
13895                            1124                            3219   
14183                            1561                            3880   
13823                            2550  

Tokelau is a territory and Niue is considered a country (As per Australian Government sources). Even though we seem to have some entities may or may not be recognized as countries, we can calculate basic statistics such as mean, minimum, maximum population. We will use the 236 entities to calculate the basic statistics.


In [ ]:
# Statistics on population
# mean, minimum, maximum in the year 2021
condition = pg_data_nona_clean["Year"]==2021
df = pg_data_nona_clean[condition]

condition = (df["Population"]==df["Population"].min())
min_country,min_population = df[condition][["Country name","Population"]].values[0]

condition = (df["Population"]==df["Population"].max())
max_country,max_population = df[condition][["Country name","Population"]].values[0]

mean_population = df["Population"].mean()

total_population = df["Population"].sum()

print(f"The minimum world population is {min_population} for {min_country}")
print(f"The maximum world population is {max_population} for country {max_country}")
print(f"The mean world population is {mean_population}.")
print(f"The total world population is {total_population}.")



The minimum world population is 1869 for Tokelau
The maximum world population is 1425893500 for country China
The mean world population is 33513966.868644066.
The total world population is 7909296181.


*In the real world, the data that is used for analysis is stored in the Gold layer*


In this case, the data contained in the variable `df` in the above cell is stored.

##### Interpret results

* Of the 236 countries analyzed, in 2021, Tokelau had the lowest world population of 1869
* In the same year, China had the highest world population of about 1.4 billion people
    * The world population ranged from 1800 to 1.4 billion people.
* Some of the countries may not be independent. These countries were also included in the analysis.
* The average world population in 2021 is about 33 million people.
* The total world population in 2021 is about 7.9 billion people.


##### Report
* The problem statement
    * A UN policy maker wants to obtain the basic statistics about the world population.
* Discuss the data selected and its source
    * Population data was used from the source United Nations, World Population Prospects (2022).
    * The data lists countries, regions (or group of countries) and their population.
    * The data also contains population across various age groups.
    * The data contained information for various years ranging from about 1950 till 2021.
* Discuss the steps used to clean the data and why
    * In this analysis, the total population across all age groups was considered.
    * The data for regions was removed from the analysis.
        * We only keep the data for various countries.
        * This resulted in 236 countries.
        * The regions we removed contain key words 'regions', 'countries', 'World' and 'UN'.
            * Including region data would cause the world population to be inaccurate.
    * A step to remove null values in the data was applied. This step did not result in any data loss. No null values were found in the data.
    * Population data discussed above was used in the analysis.  
* Discussion the analysis steps
    * What patterns did you discover?
        * Data for the year 2021 was analyzed.
        * The data consisted of countries that were islands.
            * The data pertaining to islands had the lowest population.
    * The lowest, highest, average and total population were calculated.
    * Share the interpretation of your results from above in the report
        * Of the 236 countries analyzed, in 2021, Tokelau had the lowest world population of 1869
        * In the same year, China had the highest world population of about 1.4 billion people
            * The world population ranged from 1800 to 1.4 billion people.
        * Some of the countries may not be independent. These countries were also included in the analysis.
        * The average world population in 2021 is about 33 million people.
        * The total world population in 2021 is about 7.9 billion people.
        * <*Add any figures* to your report and explain the figure.>

* Have a meeting with the requestor and discuss your findings. Provide the final data and the report to them.

Question:
Do you think the minimum and mean value skew affects the decision of the policy maker?

*In the real world, the analyzed data are typically sent to a dashboard for consumption.*



#### Summary

*Analytics Workflow*
* We saw how to ingest raw data, transform (ET of the ETL).
* We then further cleaned the data and explored it to help answer the questions.
* We discovered the minimum, maximum, total and mean population of countries in 2021.

